File: algorithms/BestTimeToBuyAndSellStock/BestTimeToBuyAndSellStockSolution.java
package BestTimeToBuyAndSellStock;

/*
 * Условие: Найти максимальную прибыль от покупки и продажи акций
 * Подход: One pass с отслеживанием минимальной цены
 * Сложность: O(n) по времени, O(1) по памяти
 */

public class BestTimeToBuyAndSellStockSolution {
    public static void main(String[] args) {
        // Test Case 1
        int[] prices1 = {7,1,5,3,6,4};
        System.out.println("Max Profit: " + maxProfit(prices1));  // 5
        
        // Test Case 2
        int[] prices2 = {7,6,4,3,1};
        System.out.println("Max Profit: " + maxProfit(prices2));  // 0
    }

    public static int maxProfit(int[] prices) {
        int minPrice = prices[0];
        int currentMaxProfit = 0;
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            }
            else if (prices[i] - minPrice > currentMaxProfit) {
                currentMaxProfit = prices[i] - minPrice;
            }
        }
        return currentMaxProfit;
    }
    
}

File: algorithms/BestTimeToBuyAndSellStock/best_time_to_buy_and_sell_stock_solution.py
"""
Условие: Найти максимальную прибыль от покупки и продажи акций
Подход: One pass с отслеживанием минимальной цены
Сложность: O(n) по времени, O(1) по памяти
"""

def maxProfit(prices):
    """
    :type prices: List[int]
    :rtype: int
    """
    min_price = prices[0]
    max_profit = 0
    for price in prices[1:]:
        if price < min_price:
            min_price = price
        elif price - min_price > max_profit:
            max_profit = price - min_price
    return max_profit


# Тесты
if __name__ == "__main__":
    # Test Case 1
    prices1 = [7,1,5,3,6,4]
    print(f"Max Profit: {maxProfit(prices1)}")  # 5
    
    # Test Case 2
    prices2 = [7,6,4,3,1]
    print(f"Max Profit: {maxProfit(prices2)}")  # 0